src/transformers/benchmark/cache_benchmark.py

```diff
@@ -0,0 +1,196 @@
+import json
+import os
+from typing import Optional
+
+import torch
+import torch._dynamo.config
+import torch._inductor.config
+from benchmark_utils_generic import BenchMark, SpeedBenchMark
+
+from transformers import AutoModelForCausalLM, AutoTokenizer, DynamicCache, StaticCache, set_seed
+
+
+os.environ["TOKENIZERS_PARALLELISM"] = "False"
+
+
+# Does multinomial sampling without a cuda synchronization
+def multinomial_sample_one_no_sync(probs_sort):
+    q = torch.empty_like(probs_sort).exponential_(1)
+    return torch.argmax(probs_sort / q, dim=-1, keepdim=True).to(dtype=torch.int)
+
+
+def logits_to_probs(logits, temperature: float = 1.0, top_k: Optional[int] = None):
+    logits = logits / max(temperature, 1e-5)
+
+    if top_k is not None:
+        v, _ = torch.topk(logits, min(top_k, logits.size(-1)))
+        pivot = v.select(-1, -1).unsqueeze(-1)
+        logits = torch.where(logits < pivot, -float("Inf"), logits)
+    probs = torch.nn.functional.softmax(logits, dim=-1)
+    return probs
+
+
+def sample(logits, temperature: float = 1.0, top_k: Optional[int] = None):
+    probs = logits_to_probs(logits[:, -1], temperature, top_k)
+    idx_next = multinomial_sample_one_no_sync(probs)
+    return idx_next, probs
+
+
+torch._inductor.config.coordinate_descent_tuning = True
+torch._inductor.config.triton.unique_kernel_names = True
+# Experimental feature to reduce compilation times, will be on by default in future
+# torch._inductor.config.fx_graph_cache = True
+
+torch.set_printoptions(linewidth=200)  # you can better see how the mask is shaped
+
+
+class CacheBenchMark(BenchMark):
+    def __init__(
+        self,
+        repo_id,
+        prefill_num_iter=3,
+        num_iter=100,
+        device="cuda",
+        attn_implementation="sdpa",
+        dtype="bfloat16",
+    ):
+        super().__init__()
+
+        self._buffer["init_kwargs"]["repo_id"] = repo_id
+        self._buffer["init_kwargs"]["prefill_num_iter"] = prefill_num_iter
+        self._buffer["init_kwargs"]["num_iter"] = num_iter
+        self._buffer["init_kwargs"]["device"] = device
+        self._buffer["init_kwargs"]["attn_implementation"] = attn_implementation
+        self._buffer["init_kwargs"]["dtype"] = dtype
+
+        self.repo_id = repo_id
+        self.prefill_num_iter = prefill_num_iter
+        self.num_iter = num_iter
+        self.device = device
+        self.dtype = dtype
+
+        dtype = getattr(torch, dtype)
+
+        token = os.getenv("HF_HUB_READ_TOKEN", None)
+        self.tokenizer = AutoTokenizer.from_pretrained(repo_id, padding_side="left", pad_token="<s>", token=token)
+        self.model = (
```

Comment:
I wonder if it would be better to pass in a model and tokenizer directly. This gives us more flexibility in how they are initialized and set. For example, what if we wanted to test padding_side="right". It means we don't need as many kwargs in the init, and we can still set values like `self.dtype` from the passed in model

Reply:
if we do this, we won't be able to save them in the configuration (report)

Reply:
Why not - can't we add them to the buffers like on L59-L64?

Reply:
passing `model` and `tokenizer` object and put in buffer is fine, but we can't save to a output file (json file).